Use std::mutex instead of std::call_once in Function when we initialize GraphExecutor.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26703 Remove convert_to_ssa argument from runCleanupPasses - it is only used in one place.
* #26705 Use optimized_graph in graph_executor.
* #26489 Use optimized graph in Inline (essentially, making Inline recursive now).
* #26488 Add 'optimized_graph' to Function.
* **#26571 Use std::mutex instead of std::call_once in Function when we initialize GraphExecutor.**
* #26487 Remove 'recurse' parameter from Inline.

We will need a mutex for computing optimized graph too, which will be
implemented in subsequent commits.

Differential Revision: [D17510883](https://our.internmc.facebook.com/intern/diff/D17510883)

Commit:
Use std::mutex instead of std::call_once in Function when we initialize GraphExecutor.

We will need a mutex for computing optimized graph too, which will be
implemented in subsequent commits.

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/function.h:
```diff
@@ -87,20 +88,27 @@ struct TORCH_API Function {
-      check_single_output();
-      executor_ = GraphExecutor(graph());
-    });
+    if (executor_initialized_) {
```
I think `if (executor_)` works here, no need for a initialized bool

Commit:
Update on "Use std::mutex instead of std::call_once in Function when we initialize GraphExecutor."


We will need a mutex for computing optimized graph too, which will be
implemented in subsequent commits.

Differential Revision: [D17510883](https://our.internmc.facebook.com/intern/diff/D17510883)

[ghstack-poisoned]

Commit:
Update on "Use std::mutex instead of std::call_once in Function when we initialize GraphExecutor."


We will need a mutex for computing optimized graph too, which will be
implemented in subsequent commits.

Differential Revision: [D17510883](https://our.internmc.facebook.com/intern/diff/D17510883)

[ghstack-poisoned]

dzhulgakov commented on torch/csrc/jit/function.h:
```diff
@@ -87,20 +88,24 @@ struct TORCH_API Function {
+    if (executor_) {
+      return executor_;
+    }
+    std::lock_guard<std::mutex> lock(compile_mutex);
```
this is a race condition if get_executor gets called from multiple threads - the read from executor_ should be guarded by the lock too
